### PR TITLE
Add compatibility file for GRUB versions up to v2.06

### PR DESCRIPTION
**Pull request by usaleem-ix**
### Motivation and Context

GRUB is not able to detect ZFS pool if snaphsot of top level boot
pool dataset is created. This issue is observed with GRUB versions
up to v2.06 if extensible_dataset feature is enabled on ZFS boot pool.

https://github.com/openzfs/zfs/issues/15261
https://github.com/openzfs/zfs/issues/13873

### Description
A new compatibility file is added `grub2-2.06` for GRUB versions
upto 2.06. `compatibility=grub2-2.06` would enable all read-only
compatible zpool features except `extensible_datase`t and other
features that depend on it.

The existing grub2 compatibility file lists all read-only features
that can be enabled on boot pool for grub with version 2.12
onwards. The issue is fixed on GRUB v2.12.

Documented both `grub2` and `grub2-2.06` files in zpool-features.7.

`livelist` feature also depends on `extensible_dataset`, but it was
documented. zpool-features.7 is updated for that as well.

### How Has This Been Tested?
Tried creating boot pool with `compatibility=grub2-2.06`, after 
creating the snapshot of top level dataset of ZFS boot pool.
Verified `grub-probe` still detects the ZFS pool.

Confirmed after re-enabling `extensible_dataset` feature and issue
occurs.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by amotin** [comment]
I am not sure this approach is really usable. Without knowing which GRUB versions are affected, it is impossible to safely choose the compatibility profile.  And the lack of understanding for the issue makes its further maintenance problematic.  Read-only-compatible features should not cause problems to GRUB, so I wonder if it is just a GRUB bug triggered by some of the features, or some feature is not so much compatible as we'd like.

**Comment by rincebrain** [comment]
It was a bug they've fixed, yes, but I don't think we're going to win the argument of making every distro cherrypick grub2 fixes in a timely fashion, compared to people running our newest versions or people packaging newer things than the distro does, and getting burned.

I said in another one, I think #15459, that I think we should do the opposite, and make `grub2` the minimum set of features we know don't break grub2 anywhere, since that's historically what we used this for. I realize this makes people dirty retroactively, which is unfortunate, but that's a tradeoff of historical pools causing "why is it saying this" versus potentially unbounded versions of new pools causing surprises if they follow what we used the `compatibility=grub2` option to mean historically.

Admittedly, this is biased by me not looking forward to having to explain to a lot of people who will come by asking why `-o compatibility=grub2` broke boot for them until the unfixed versions of grub2 are entirely gone from the list of things people commonly install, but I think I'd make the same argument even if I wasn't going to be the one often explaining this to people later.

**Comment by usaleem-ix** [comment]
I have updated the PR and added a new compatibility file, `grub2-2.06` for GRUB versions up to 2.06. `compatibility=grub2-2.06` would enable all read-only compatible zpool features except `extensible_dataset` and other features that depend on it.

The existing grub2 compatibility file lists all read-only features that can be enabled on boot pool for grub with version 2.12
onwards.

**Review by amotin — APPROVED** [review]

File: tests/zfs-tests/tests/functional/cli_root/zpool_create/zpool_create.shlib [hunk 1]
```diff
@@ -119,9 +119,8 @@ function check_feature_set
 	while read line; do
 		typeset flag=1
 
-		if [[ "$line" == "#*" ]]; then
-			continue
-		fi
+		# Skip comments
+		[[ $line = \#* ]] && continue
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zpool_create/zpool_create.shlib:123]: This seemed entirely unrelated and belongs in it's own commit.

File: cmd/zpool/compatibility.d/grub2-2.06 [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Features which are supported by GRUB2 versions prior to v2.12.
```
Comment (behlendorf) [inline cmd/zpool/compatibility.d/grub2-2.06:1]: Having this compatibility file named `grub2-2.06` yet reference v2.12 as the cut off is a bit confusing.  Is this because most/all distributions jumped from v2.06 to v2.12 and there effectively aren't versions in between out in the wild?

Comment (amotin) [reply cmd/zpool/compatibility.d/grub2-2.06:1]: I don't see a confusion in specifying the oldest version supported by respective file.  I think when somebody chooses among different versions, he should be conservative and select biggest available version smaller or equal to his requirements.  This file supports versions [2.06, 2.12).  The confusion here is that version supporting 2.12 and up is called just "grub2".

Comment (behlendorf) [reply cmd/zpool/compatibility.d/grub2-2.06:1]: Yes, that's unfortunate.  Rather than live with this confusion forever what if we renamed the "grub2" file to "grub2-2.12" for the 2.3 release.  Then put in place a "grub2" -> "grub2-2.12" symlink.  We could also backport this change to the 2.2 releases.  Once in place we could consider retiring the "grub2" file in a future release.  Users would then need to specifically set "compatibility=grub2-2.12" if they were using the "grub2" config.

Comment (amotin) [reply cmd/zpool/compatibility.d/grub2-2.06:1]: Yea, something like that could probably be done.  I just don't remember what is the reaction of zpool on absent compatibility file.  We could look on how to make it less annoying for user and still notify that intervention is required.

Comment (behlendorf) [reply cmd/zpool/compatibility.d/grub2-2.06:1]: That sounds good to me.  We don't even need to be in a rush to remove the "grub2" symlink.  Although, I agree it'd be good to have some notification/warning that encourage users to update their `compatibility` property.  We'd also want to discourage new pools from being created using the "grub2" symlink.

File: tests/zfs-tests/tests/functional/cli_root/zpool_create/zpool_create.shlib (hunk 1 above)
Comment (usaleem-ix) [reply tests/zfs-tests/tests/functional/cli_root/zpool_create/zpool_create.shlib:123]: I have created a separate commit for this.

File: cmd/zpool/compatibility.d/grub2-2.06 (hunk 2 above)
Comment (usaleem-ix) [reply cmd/zpool/compatibility.d/grub2-2.06:1]: I have renamed the `grub2` compatibility file to `grub2-2.12` since it contains the features set that is compatible with grub2 version 2.12 onwards.

A new symlink named `grub2` is created that points to `grub2-2.12` file.

File: tests/zfs-tests/tests/functional/cli_root/zpool_create/zpool_create.shlib (hunk 1 above)
Comment (behlendorf) [reply tests/zfs-tests/tests/functional/cli_root/zpool_create/zpool_create.shlib:123]: That clears things up.  Thanks.

File: cmd/zpool/Makefile.am [hunk 3]
```diff
@@ -141,6 +141,8 @@ dist_zpoolcompat_DATA = \
 	%D%/compatibility.d/freebsd-11.3 \
 	%D%/compatibility.d/freenas-9.10.2 \
 	%D%/compatibility.d/grub2 \
```
[the remark this answers is not in the document]
Comment (usaleem-ix) [reply cmd/zpool/Makefile.am]: Thanks, I have removed the symlink and updated accordingly.

**Review by behlendorf — APPROVED** [review]
